// hpack.js/tools/dist/gen-static-table.dev.js
"use strict";

var utils = require('./utils');

var table = "\n1\t:authority\t\n2\t:method\tGET\n3\t:method\tPOST\n4\t:path\t/\n5\t:path\t/index.html\n6\t:scheme\thttp\n7\t:scheme\thttps\n8\t:status\t200\n9\t:status\t204\n10\t:status\t206\n11\t:status\t304\n12\t:status\t400\n13\t:status\t404\n14\t:status\t500\n15\taccept-charset\t\n16\taccept-encoding\tgzip, deflate\n17\taccept-language\t\n18\taccept-ranges\t\n19\taccept\t\n20\taccess-control-allow-origin\t\n21\tage\t\n22\tallow\t\n23\tauthorization\t\n24\tcache-control\t\n25\tcontent-disposition\t\n26\tcontent-encoding\t\n27\tcontent-language\t\n28\tcontent-length\t\n29\tcontent-location\t\n30\tcontent-range\t\n31\tcontent-type\t\n32\tcookie\t\n33\tdate\t\n34\tetag\t\n35\texpect\t\n36\texpires\t\n37\tfrom\t\n38\thost\t\n39\tif-match\t\n40\tif-modified-since\t\n41\tif-none-match\t\n42\tif-range\t\n43\tif-unmodified-since\t\n44\tlast-modified\t\n45\tlink\t\n46\tlocation\t\n47\tmax-forwards\t\n48\tproxy-authenticate\t\n49\tproxy-authorization\t\n50\trange\t\n51\treferer\t\n52\trefresh\t\n53\tretry-after\t\n54\tserver\t\n55\tset-cookie\t\n56\tstrict-transport-security\t\n57\ttransfer-encoding\t\n58\tuser-agent\t\n59\tvary\t\n60\tvia\t\n61\twww-authenticate\t\n";
var out = [];
table.split('\n').filter(function (line) {
  return line;
}).forEach(function (line) {
  var columns = line.split(/\t/g);
  var name = columns[1];
  var value = columns[2];
  var nameSize = Buffer.byteLength(name);
  var valueSize = Buffer.byteLength(value);
  out.push({
    name: name,
    value: value,
    nameSize: nameSize,
    totalSize: nameSize + valueSize + 32
  });
});
console.log('exports.table = ' + JSON.stringify(out, false, 2) + ';');
var map = {};
table.split('\n').filter(function (line) {
  return line;
}).forEach(function (line) {
  var columns = line.split(/\t/g);
  var name = columns[1];
  var value = columns[2];
  var index = columns[0] | 0;

  if (!map[name]) {
    map[name] = {
      index: index,
      values: {}
    };
  }

  map[name].values[value] = index;
});
console.log('exports.map = ' + JSON.stringify(map, false, 2) + ';');
