// yargs/build/lib/dist/command.dev.js
"use strict";

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _toArray(arr) { return _arrayWithHoles(arr) || _iterableToArray(arr) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance"); }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

function _toConsumableArray(arr) { return _arrayWithoutHoles(arr) || _iterableToArray(arr) || _nonIterableSpread(); }

function _nonIterableSpread() { throw new TypeError("Invalid attempt to spread non-iterable instance"); }

function _iterableToArray(iter) { if (Symbol.iterator in Object(iter) || Object.prototype.toString.call(iter) === "[object Arguments]") return Array.from(iter); }

function _arrayWithoutHoles(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = new Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.isCommandBuilderCallback = exports.isCommandBuilderDefinition = exports.isCommandHandlerDefinition = exports.command = void 0;

var common_types_1 = require("./common-types");

var is_promise_1 = require("./is-promise");

var middleware_1 = require("./middleware");

var parse_command_1 = require("./parse-command");

var path = require("path");

var util_1 = require("util");

var yargs_1 = require("./yargs");

var requireDirectory = require("require-directory");

var whichModule = require("which-module");

var Parser = require("yargs-parser");

var DEFAULT_MARKER = /(^\*)|(^\$0)/; // handles parsing positional arguments,
// and populating argv with said positional
// arguments.

function command(yargs, usage, validation) {
  var globalMiddleware = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : [];
  var self = {};
  var handlers = {};
  var aliasMap = {};
  var defaultCommand;

  self.addHandler = function addHandler(cmd, description, builder, handler, commandMiddleware, deprecated) {
    var aliases = [];
    var middlewares = middleware_1.commandMiddlewareFactory(commandMiddleware);

    handler = handler || function () {};

    if (Array.isArray(cmd)) {
      aliases = cmd.slice(1);
      cmd = cmd[0];
    } else if (isCommandHandlerDefinition(cmd)) {
      var _command = Array.isArray(cmd.command) || typeof cmd.command === 'string' ? cmd.command : moduleName(cmd);

      if (cmd.aliases) _command = [].concat(_command).concat(cmd.aliases);
      self.addHandler(_command, extractDesc(cmd), cmd.builder, cmd.handler, cmd.middlewares, cmd.deprecated);
      return;
    } // allow a module to be provided instead of separate builder and handler


    if (isCommandBuilderDefinition(builder)) {
      self.addHandler([cmd].concat(aliases), description, builder.builder, builder.handler, builder.middlewares, builder.deprecated);
      return;
    } // parse positionals out of cmd string


    var parsedCommand = parse_command_1.parseCommand(cmd); // remove positional args from aliases only

    aliases = aliases.map(function (alias) {
      return parse_command_1.parseCommand(alias).cmd;
    }); // check for default and filter out '*''

    var isDefault = false;
    var parsedAliases = [parsedCommand.cmd].concat(aliases).filter(function (c) {
      if (DEFAULT_MARKER.test(c)) {
        isDefault = true;
        return false;
      }

      return true;
    }); // standardize on $0 for default command.

    if (parsedAliases.length === 0 && isDefault) parsedAliases.push('$0'); // shift cmd and aliases after filtering out '*'

    if (isDefault) {
      parsedCommand.cmd = parsedAliases[0];
      aliases = parsedAliases.slice(1);
      cmd = cmd.replace(DEFAULT_MARKER, parsedCommand.cmd);
    } // populate aliasMap


    aliases.forEach(function (alias) {
      aliasMap[alias] = parsedCommand.cmd;
    });

    if (description !== false) {
      usage.command(cmd, description, isDefault, aliases, deprecated);
    }

    handlers[parsedCommand.cmd] = {
      original: cmd,
      description: description,
      handler: handler,
      builder: builder || {},
      middlewares: middlewares,
      deprecated: deprecated,
      demanded: parsedCommand.demanded,
      optional: parsedCommand.optional
    };
    if (isDefault) defaultCommand = handlers[parsedCommand.cmd];
  };

  self.addDirectory = function addDirectory(dir, context, req, callerFile, opts) {
    opts = opts || {}; // disable recursion to support nested directories of subcommands

    if (typeof opts.recurse !== 'boolean') opts.recurse = false; // exclude 'json', 'coffee' from require-directory defaults

    if (!Array.isArray(opts.extensions)) opts.extensions = ['js']; // allow consumer to define their own visitor function

    var parentVisit = typeof opts.visit === 'function' ? opts.visit : function (o) {
      return o;
    }; // call addHandler via visitor function

    opts.visit = function visit(obj, joined, filename) {
      var visited = parentVisit(obj, joined, filename); // allow consumer to skip modules with their own visitor

      if (visited) {
        // check for cyclic reference
        // each command file path should only be seen once per execution
        if (~context.files.indexOf(joined)) return visited; // keep track of visited files in context.files

        context.files.push(joined);
        self.addHandler(visited);
      }

      return visited;
    };

    requireDirectory({
      require: req,
      filename: callerFile
    }, dir, opts);
  }; // lookup module object from require()d command and derive name
  // if module was not require()d and no name given, throw error


  function moduleName(obj) {
    var mod = whichModule(obj);
    if (!mod) throw new Error("No command name given for module: ".concat(util_1.inspect(obj)));
    return commandFromFilename(mod.filename);
  } // derive command name from filename


  function commandFromFilename(filename) {
    return path.basename(filename, path.extname(filename));
  }

  function extractDesc(_ref) {
    var describe = _ref.describe,
        description = _ref.description,
        desc = _ref.desc;

    for (var _i = 0, _arr = [describe, description, desc]; _i < _arr.length; _i++) {
      var test = _arr[_i];
      if (typeof test === 'string' || test === false) return test;
      common_types_1.assertNotStrictEqual(test, true);
    }

    return false;
  }

  self.getCommands = function () {
    return Object.keys(handlers).concat(Object.keys(aliasMap));
  };

  self.getCommandHandlers = function () {
    return handlers;
  };

  self.hasDefaultCommand = function () {
    return !!defaultCommand;
  };

  self.runCommand = function runCommand(command, yargs, parsed, commandIndex) {
    var aliases = parsed.aliases;
    var commandHandler = handlers[command] || handlers[aliasMap[command]] || defaultCommand;
    var currentContext = yargs.getContext();
    var numFiles = currentContext.files.length;
    var parentCommands = currentContext.commands.slice(); // what does yargs look like after the builder is run?

    var innerArgv = parsed.argv;
    var positionalMap = {};

    if (command) {
      currentContext.commands.push(command);
      currentContext.fullCommands.push(commandHandler.original);
    }

    var builder = commandHandler.builder;

    if (isCommandBuilderCallback(builder)) {
      // a function can be provided, which builds
      // up a yargs chain and possibly returns it.
      var builderOutput = builder(yargs.reset(parsed.aliases));
      var innerYargs = yargs_1.isYargsInstance(builderOutput) ? builderOutput : yargs;

      if (shouldUpdateUsage(innerYargs)) {
        innerYargs.getUsageInstance().usage(usageFromParentCommandsCommandHandler(parentCommands, commandHandler), commandHandler.description);
      }

      innerArgv = innerYargs._parseArgs(null, null, true, commandIndex);
      aliases = innerYargs.parsed.aliases;
    } else if (isCommandBuilderOptionDefinitions(builder)) {
      // as a short hand, an object can instead be provided, specifying
      // the options that a command takes.
      var _innerYargs = yargs.reset(parsed.aliases);

      if (shouldUpdateUsage(_innerYargs)) {
        _innerYargs.getUsageInstance().usage(usageFromParentCommandsCommandHandler(parentCommands, commandHandler), commandHandler.description);
      }

      Object.keys(commandHandler.builder).forEach(function (key) {
        _innerYargs.option(key, builder[key]);
      });
      innerArgv = _innerYargs._parseArgs(null, null, true, commandIndex);
      aliases = _innerYargs.parsed.aliases;
    }

    if (!yargs._hasOutput()) {
      positionalMap = populatePositionals(commandHandler, innerArgv, currentContext);
    }

    var middlewares = globalMiddleware.slice(0).concat(commandHandler.middlewares);
    middleware_1.applyMiddleware(innerArgv, yargs, middlewares, true); // we apply validation post-hoc, so that custom
    // checks get passed populated positional arguments.

    if (!yargs._hasOutput()) {
      yargs._runValidation(innerArgv, aliases, positionalMap, yargs.parsed.error, !command);
    }

    if (commandHandler.handler && !yargs._hasOutput()) {
      yargs._setHasOutput(); // to simplify the parsing of positionals in commands,
      // we temporarily populate '--' rather than _, with arguments


      var populateDoubleDash = !!yargs.getOptions().configuration['populate--'];
      if (!populateDoubleDash) yargs._copyDoubleDash(innerArgv);
      innerArgv = middleware_1.applyMiddleware(innerArgv, yargs, middlewares, false);
      var handlerResult;

      if (is_promise_1.isPromise(innerArgv)) {
        handlerResult = innerArgv.then(function (argv) {
          return commandHandler.handler(argv);
        });
      } else {
        handlerResult = commandHandler.handler(innerArgv);
      }

      var handlerFinishCommand = yargs.getHandlerFinishCommand();

      if (is_promise_1.isPromise(handlerResult)) {
        yargs.getUsageInstance().cacheHelpMessage();
        handlerResult.then(function (value) {
          if (handlerFinishCommand) {
            handlerFinishCommand(value);
          }
        })["catch"](function (error) {
          try {
            yargs.getUsageInstance().fail(null, error);
          } catch (err) {// fail's throwing would cause an unhandled rejection.
          }
        }).then(function () {
          yargs.getUsageInstance().clearCachedHelpMessage();
        });
      } else {
        if (handlerFinishCommand) {
          handlerFinishCommand(handlerResult);
        }
      }
    }

    if (command) {
      currentContext.commands.pop();
      currentContext.fullCommands.pop();
    }

    numFiles = currentContext.files.length - numFiles;
    if (numFiles > 0) currentContext.files.splice(numFiles * -1, numFiles);
    return innerArgv;
  };

  function shouldUpdateUsage(yargs) {
    return !yargs.getUsageInstance().getUsageDisabled() && yargs.getUsageInstance().getUsage().length === 0;
  }

  function usageFromParentCommandsCommandHandler(parentCommands, commandHandler) {
    var c = DEFAULT_MARKER.test(commandHandler.original) ? commandHandler.original.replace(DEFAULT_MARKER, '').trim() : commandHandler.original;
    var pc = parentCommands.filter(function (c) {
      return !DEFAULT_MARKER.test(c);
    });
    pc.push(c);
    return "$0 ".concat(pc.join(' '));
  }

  self.runDefaultBuilderOn = function (yargs) {
    common_types_1.assertNotStrictEqual(defaultCommand, undefined);

    if (shouldUpdateUsage(yargs)) {
      // build the root-level command string from the default string.
      var commandString = DEFAULT_MARKER.test(defaultCommand.original) ? defaultCommand.original : defaultCommand.original.replace(/^[^[\]<>]*/, '$0 ');
      yargs.getUsageInstance().usage(commandString, defaultCommand.description);
    }

    var builder = defaultCommand.builder;

    if (isCommandBuilderCallback(builder)) {
      builder(yargs);
    } else {
      Object.keys(builder).forEach(function (key) {
        yargs.option(key, builder[key]);
      });
    }
  }; // transcribe all positional arguments "command <foo> <bar> [apple]"
  // onto argv.


  function populatePositionals(commandHandler, argv, context) {
    argv._ = argv._.slice(context.commands.length); // nuke the current commands

    var demanded = commandHandler.demanded.slice(0);
    var optional = commandHandler.optional.slice(0);
    var positionalMap = {};
    validation.positionalCount(demanded.length, argv._.length);

    while (demanded.length) {
      var demand = demanded.shift();
      populatePositional(demand, argv, positionalMap);
    }

    while (optional.length) {
      var maybe = optional.shift();
      populatePositional(maybe, argv, positionalMap);
    }

    argv._ = context.commands.concat(argv._);
    postProcessPositionals(argv, positionalMap, self.cmdToParseOptions(commandHandler.original));
    return positionalMap;
  }

  function populatePositional(positional, argv, positionalMap) {
    var cmd = positional.cmd[0];

    if (positional.variadic) {
      positionalMap[cmd] = argv._.splice(0).map(String);
    } else {
      if (argv._.length) positionalMap[cmd] = [String(argv._.shift())];
    }
  } // we run yargs-parser against the positional arguments
  // applying the same parsing logic used for flags.


  function postProcessPositionals(argv, positionalMap, parseOptions) {
    // combine the parsing hints we've inferred from the command
    // string with explicitly configured parsing hints.
    var options = Object.assign({}, yargs.getOptions());
    options["default"] = Object.assign(parseOptions["default"], options["default"]);

    for (var _i2 = 0, _Object$keys = Object.keys(parseOptions.alias); _i2 < _Object$keys.length; _i2++) {
      var key = _Object$keys[_i2];
      options.alias[key] = (options.alias[key] || []).concat(parseOptions.alias[key]);
    }

    options.array = options.array.concat(parseOptions.array);
    delete options.config; //  don't load config when processing positionals.

    var unparsed = [];
    Object.keys(positionalMap).forEach(function (key) {
      positionalMap[key].map(function (value) {
        if (options.configuration['unknown-options-as-args']) options.key[key] = true;
        unparsed.push("--".concat(key));
        unparsed.push(value);
      });
    }); // short-circuit parse.

    if (!unparsed.length) return;
    var config = Object.assign({}, options.configuration, {
      'populate--': true
    });
    var parsed = Parser.detailed(unparsed, Object.assign({}, options, {
      configuration: config
    }));

    if (parsed.error) {
      yargs.getUsageInstance().fail(parsed.error.message, parsed.error);
    } else {
      // only copy over positional keys (don't overwrite
      // flag arguments that were already parsed).
      var positionalKeys = Object.keys(positionalMap);
      Object.keys(positionalMap).forEach(function (key) {
        positionalKeys.push.apply(positionalKeys, _toConsumableArray(parsed.aliases[key]));
      });
      Object.keys(parsed.argv).forEach(function (key) {
        if (positionalKeys.indexOf(key) !== -1) {
          // any new aliases need to be placed in positionalMap, which
          // is used for validation.
          if (!positionalMap[key]) positionalMap[key] = parsed.argv[key];
          argv[key] = parsed.argv[key];
        }
      });
    }
  }

  self.cmdToParseOptions = function (cmdString) {
    var parseOptions = {
      array: [],
      "default": {},
      alias: {},
      demand: {}
    };
    var parsed = parse_command_1.parseCommand(cmdString);
    parsed.demanded.forEach(function (d) {
      var _d$cmd = _toArray(d.cmd),
          cmd = _d$cmd[0],
          aliases = _d$cmd.slice(1);

      if (d.variadic) {
        parseOptions.array.push(cmd);
        parseOptions["default"][cmd] = [];
      }

      parseOptions.alias[cmd] = aliases;
      parseOptions.demand[cmd] = true;
    });
    parsed.optional.forEach(function (o) {
      var _o$cmd = _toArray(o.cmd),
          cmd = _o$cmd[0],
          aliases = _o$cmd.slice(1);

      if (o.variadic) {
        parseOptions.array.push(cmd);
        parseOptions["default"][cmd] = [];
      }

      parseOptions.alias[cmd] = aliases;
    });
    return parseOptions;
  };

  self.reset = function () {
    handlers = {};
    aliasMap = {};
    defaultCommand = undefined;
    return self;
  }; // used by yargs.parse() to freeze
  // the state of commands such that
  // we can apply .parse() multiple times
  // with the same yargs instance.


  var frozens = [];

  self.freeze = function () {
    frozens.push({
      handlers: handlers,
      aliasMap: aliasMap,
      defaultCommand: defaultCommand
    });
  };

  self.unfreeze = function () {
    var frozen = frozens.pop();
    common_types_1.assertNotStrictEqual(frozen, undefined);
    handlers = frozen.handlers;
    aliasMap = frozen.aliasMap;
    defaultCommand = frozen.defaultCommand;
  };

  return self;
}

exports.command = command;

function isCommandHandlerDefinition(cmd) {
  return _typeof(cmd) === 'object';
}

exports.isCommandHandlerDefinition = isCommandHandlerDefinition;

function isCommandBuilderDefinition(builder) {
  return _typeof(builder) === 'object' && !!builder.builder && typeof builder.handler === 'function';
}

exports.isCommandBuilderDefinition = isCommandBuilderDefinition;

function isCommandBuilderCallback(builder) {
  return typeof builder === 'function';
}

exports.isCommandBuilderCallback = isCommandBuilderCallback;

function isCommandBuilderOptionDefinitions(builder) {
  return _typeof(builder) === 'object';
}
