// vue/src/platforms/web/dist/entry-compiler.dev.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "parseComponent", {
  enumerable: true,
  get: function get() {
    return _parser.parseComponent;
  }
});
Object.defineProperty(exports, "compile", {
  enumerable: true,
  get: function get() {
    return _index.compile;
  }
});
Object.defineProperty(exports, "compileToFunctions", {
  enumerable: true,
  get: function get() {
    return _index.compileToFunctions;
  }
});
Object.defineProperty(exports, "ssrCompile", {
  enumerable: true,
  get: function get() {
    return _compiler.ssrCompile;
  }
});
Object.defineProperty(exports, "ssrCompileToFunctions", {
  enumerable: true,
  get: function get() {
    return _compiler.ssrCompileToFunctions;
  }
});
Object.defineProperty(exports, "generateCodeFrame", {
  enumerable: true,
  get: function get() {
    return _codeframe.generateCodeFrame;
  }
});

var _parser = require("sfc/parser");

var _index = require("./compiler/index");

var _compiler = require("./server/compiler");

var _codeframe = require("compiler/codeframe");
