// eslint-scope/lib/dist/scope-manager.dev.js
/*
  Copyright (C) 2015 Yusuke Suzuki <utatane.tea@gmail.com>

  Redistribution and use in source and binary forms, with or without
  modification, are permitted provided that the following conditions are met:

    * Redistributions of source code must retain the above copyright
      notice, this list of conditions and the following disclaimer.
    * Redistributions in binary form must reproduce the above copyright
      notice, this list of conditions and the following disclaimer in the
      documentation and/or other materials provided with the distribution.

  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
  ARE DISCLAIMED. IN NO EVENT SHALL <COPYRIGHT HOLDER> BE LIABLE FOR ANY
  DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
  (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
  ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
  (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
  THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
*/
"use strict";
/* eslint-disable no-underscore-dangle */

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

var Scope = require("./scope");

var assert = require("assert");

var GlobalScope = Scope.GlobalScope;
var CatchScope = Scope.CatchScope;
var WithScope = Scope.WithScope;
var ModuleScope = Scope.ModuleScope;
var ClassScope = Scope.ClassScope;
var SwitchScope = Scope.SwitchScope;
var FunctionScope = Scope.FunctionScope;
var ForScope = Scope.ForScope;
var FunctionExpressionNameScope = Scope.FunctionExpressionNameScope;
var BlockScope = Scope.BlockScope;
/**
 * @class ScopeManager
 */

var ScopeManager =
/*#__PURE__*/
function () {
  function ScopeManager(options) {
    _classCallCheck(this, ScopeManager);

    this.scopes = [];
    this.globalScope = null;
    this.__nodeToScope = new WeakMap();
    this.__currentScope = null;
    this.__options = options;
    this.__declaredVariables = new WeakMap();
  }

  _createClass(ScopeManager, [{
    key: "__useDirective",
    value: function __useDirective() {
      return this.__options.directive;
    }
  }, {
    key: "__isOptimistic",
    value: function __isOptimistic() {
      return this.__options.optimistic;
    }
  }, {
    key: "__ignoreEval",
    value: function __ignoreEval() {
      return this.__options.ignoreEval;
    }
  }, {
    key: "__isNodejsScope",
    value: function __isNodejsScope() {
      return this.__options.nodejsScope;
    }
  }, {
    key: "isModule",
    value: function isModule() {
      return this.__options.sourceType === "module";
    }
  }, {
    key: "isImpliedStrict",
    value: function isImpliedStrict() {
      return this.__options.impliedStrict;
    }
  }, {
    key: "isStrictModeSupported",
    value: function isStrictModeSupported() {
      return this.__options.ecmaVersion >= 5;
    } // Returns appropriate scope for this node.

  }, {
    key: "__get",
    value: function __get(node) {
      return this.__nodeToScope.get(node);
    }
    /**
     * Get variables that are declared by the node.
     *
     * "are declared by the node" means the node is same as `Variable.defs[].node` or `Variable.defs[].parent`.
     * If the node declares nothing, this method returns an empty array.
     * CAUTION: This API is experimental. See https://github.com/estools/escope/pull/69 for more details.
     *
     * @param {Espree.Node} node - a node to get.
     * @returns {Variable[]} variables that declared by the node.
     */

  }, {
    key: "getDeclaredVariables",
    value: function getDeclaredVariables(node) {
      return this.__declaredVariables.get(node) || [];
    }
    /**
     * acquire scope from node.
     * @method ScopeManager#acquire
     * @param {Espree.Node} node - node for the acquired scope.
     * @param {boolean=} inner - look up the most inner scope, default value is false.
     * @returns {Scope?} Scope from node
     */

  }, {
    key: "acquire",
    value: function acquire(node, inner) {
      /**
       * predicate
       * @param {Scope} testScope - scope to test
       * @returns {boolean} predicate
       */
      function predicate(testScope) {
        if (testScope.type === "function" && testScope.functionExpressionScope) {
          return false;
        }

        return true;
      }

      var scopes = this.__get(node);

      if (!scopes || scopes.length === 0) {
        return null;
      } // Heuristic selection from all scopes.
      // If you would like to get all scopes, please use ScopeManager#acquireAll.


      if (scopes.length === 1) {
        return scopes[0];
      }

      if (inner) {
        for (var i = scopes.length - 1; i >= 0; --i) {
          var scope = scopes[i];

          if (predicate(scope)) {
            return scope;
          }
        }
      } else {
        for (var _i = 0, iz = scopes.length; _i < iz; ++_i) {
          var _scope = scopes[_i];

          if (predicate(_scope)) {
            return _scope;
          }
        }
      }

      return null;
    }
    /**
     * acquire all scopes from node.
     * @method ScopeManager#acquireAll
     * @param {Espree.Node} node - node for the acquired scope.
     * @returns {Scopes?} Scope array
     */

  }, {
    key: "acquireAll",
    value: function acquireAll(node) {
      return this.__get(node);
    }
    /**
     * release the node.
     * @method ScopeManager#release
     * @param {Espree.Node} node - releasing node.
     * @param {boolean=} inner - look up the most inner scope, default value is false.
     * @returns {Scope?} upper scope for the node.
     */

  }, {
    key: "release",
    value: function release(node, inner) {
      var scopes = this.__get(node);

      if (scopes && scopes.length) {
        var scope = scopes[0].upper;

        if (!scope) {
          return null;
        }

        return this.acquire(scope.block, inner);
      }

      return null;
    }
  }, {
    key: "attach",
    value: function attach() {} // eslint-disable-line class-methods-use-this

  }, {
    key: "detach",
    value: function detach() {} // eslint-disable-line class-methods-use-this

  }, {
    key: "__nestScope",
    value: function __nestScope(scope) {
      if (scope instanceof GlobalScope) {
        assert(this.__currentScope === null);
        this.globalScope = scope;
      }

      this.__currentScope = scope;
      return scope;
    }
  }, {
    key: "__nestGlobalScope",
    value: function __nestGlobalScope(node) {
      return this.__nestScope(new GlobalScope(this, node));
    }
  }, {
    key: "__nestBlockScope",
    value: function __nestBlockScope(node) {
      return this.__nestScope(new BlockScope(this, this.__currentScope, node));
    }
  }, {
    key: "__nestFunctionScope",
    value: function __nestFunctionScope(node, isMethodDefinition) {
      return this.__nestScope(new FunctionScope(this, this.__currentScope, node, isMethodDefinition));
    }
  }, {
    key: "__nestForScope",
    value: function __nestForScope(node) {
      return this.__nestScope(new ForScope(this, this.__currentScope, node));
    }
  }, {
    key: "__nestCatchScope",
    value: function __nestCatchScope(node) {
      return this.__nestScope(new CatchScope(this, this.__currentScope, node));
    }
  }, {
    key: "__nestWithScope",
    value: function __nestWithScope(node) {
      return this.__nestScope(new WithScope(this, this.__currentScope, node));
    }
  }, {
    key: "__nestClassScope",
    value: function __nestClassScope(node) {
      return this.__nestScope(new ClassScope(this, this.__currentScope, node));
    }
  }, {
    key: "__nestSwitchScope",
    value: function __nestSwitchScope(node) {
      return this.__nestScope(new SwitchScope(this, this.__currentScope, node));
    }
  }, {
    key: "__nestModuleScope",
    value: function __nestModuleScope(node) {
      return this.__nestScope(new ModuleScope(this, this.__currentScope, node));
    }
  }, {
    key: "__nestFunctionExpressionNameScope",
    value: function __nestFunctionExpressionNameScope(node) {
      return this.__nestScope(new FunctionExpressionNameScope(this, this.__currentScope, node));
    }
  }, {
    key: "__isES6",
    value: function __isES6() {
      return this.__options.ecmaVersion >= 6;
    }
  }]);

  return ScopeManager;
}();

module.exports = ScopeManager;
/* vim: set sw=4 ts=4 et tw=80 : */
