// eslint-scope/lib/dist/scope.dev.js
/*
  Copyright (C) 2015 Yusuke Suzuki <utatane.tea@gmail.com>

  Redistribution and use in source and binary forms, with or without
  modification, are permitted provided that the following conditions are met:

    * Redistributions of source code must retain the above copyright
      notice, this list of conditions and the following disclaimer.
    * Redistributions in binary form must reproduce the above copyright
      notice, this list of conditions and the following disclaimer in the
      documentation and/or other materials provided with the distribution.

  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
  ARE DISCLAIMED. IN NO EVENT SHALL <COPYRIGHT HOLDER> BE LIABLE FOR ANY
  DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
  (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
  ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
  (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
  THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
*/
"use strict";
/* eslint-disable no-underscore-dangle */

/* eslint-disable no-undefined */

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _get(target, property, receiver) { if (typeof Reflect !== "undefined" && Reflect.get) { _get = Reflect.get; } else { _get = function _get(target, property, receiver) { var base = _superPropBase(target, property); if (!base) return; var desc = Object.getOwnPropertyDescriptor(base, property); if (desc.get) { return desc.get.call(receiver); } return desc.value; }; } return _get(target, property, receiver || target); }

function _superPropBase(object, property) { while (!Object.prototype.hasOwnProperty.call(object, property)) { object = _getPrototypeOf(object); if (object === null) break; } return object; }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

var Syntax = require("estraverse").Syntax;

var Reference = require("./reference");

var Variable = require("./variable");

var Definition = require("./definition").Definition;

var assert = require("assert");
/**
 * Test if scope is struct
 * @param {Scope} scope - scope
 * @param {Block} block - block
 * @param {boolean} isMethodDefinition - is method definiton
 * @param {boolean} useDirective - use directive
 * @returns {boolean} is strict scope
 */


function isStrictScope(scope, block, isMethodDefinition, useDirective) {
  var body; // When upper scope is exists and strict, inner scope is also strict.

  if (scope.upper && scope.upper.isStrict) {
    return true;
  }

  if (isMethodDefinition) {
    return true;
  }

  if (scope.type === "class" || scope.type === "module") {
    return true;
  }

  if (scope.type === "block" || scope.type === "switch") {
    return false;
  }

  if (scope.type === "function") {
    if (block.type === Syntax.ArrowFunctionExpression && block.body.type !== Syntax.BlockStatement) {
      return false;
    }

    if (block.type === Syntax.Program) {
      body = block;
    } else {
      body = block.body;
    }

    if (!body) {
      return false;
    }
  } else if (scope.type === "global") {
    body = block;
  } else {
    return false;
  } // Search 'use strict' directive.


  if (useDirective) {
    for (var i = 0, iz = body.body.length; i < iz; ++i) {
      var stmt = body.body[i];

      if (stmt.type !== Syntax.DirectiveStatement) {
        break;
      }

      if (stmt.raw === "\"use strict\"" || stmt.raw === "'use strict'") {
        return true;
      }
    }
  } else {
    for (var _i = 0, _iz = body.body.length; _i < _iz; ++_i) {
      var _stmt = body.body[_i];

      if (_stmt.type !== Syntax.ExpressionStatement) {
        break;
      }

      var expr = _stmt.expression;

      if (expr.type !== Syntax.Literal || typeof expr.value !== "string") {
        break;
      }

      if (expr.raw !== null && expr.raw !== undefined) {
        if (expr.raw === "\"use strict\"" || expr.raw === "'use strict'") {
          return true;
        }
      } else {
        if (expr.value === "use strict") {
          return true;
        }
      }
    }
  }

  return false;
}
/**
 * Register scope
 * @param {ScopeManager} scopeManager - scope manager
 * @param {Scope} scope - scope
 * @returns {void}
 */


function registerScope(scopeManager, scope) {
  scopeManager.scopes.push(scope);

  var scopes = scopeManager.__nodeToScope.get(scope.block);

  if (scopes) {
    scopes.push(scope);
  } else {
    scopeManager.__nodeToScope.set(scope.block, [scope]);
  }
}
/**
 * Should be statically
 * @param {Object} def - def
 * @returns {boolean} should be statically
 */


function shouldBeStatically(def) {
  return def.type === Variable.ClassName || def.type === Variable.Variable && def.parent.kind !== "var";
}
/**
 * @class Scope
 */


var Scope =
/*#__PURE__*/
function () {
  function Scope(scopeManager, type, upperScope, block, isMethodDefinition) {
    _classCallCheck(this, Scope);

    /**
     * One of 'module', 'block', 'switch', 'function', 'catch', 'with', 'function', 'class', 'global'.
     * @member {String} Scope#type
     */
    this.type = type;
    /**
    * The scoped {@link Variable}s of this scope, as <code>{ Variable.name
    * : Variable }</code>.
    * @member {Map} Scope#set
    */

    this.set = new Map();
    /**
     * The tainted variables of this scope, as <code>{ Variable.name :
     * boolean }</code>.
     * @member {Map} Scope#taints */

    this.taints = new Map();
    /**
     * Generally, through the lexical scoping of JS you can always know
     * which variable an identifier in the source code refers to. There are
     * a few exceptions to this rule. With 'global' and 'with' scopes you
     * can only decide at runtime which variable a reference refers to.
     * Moreover, if 'eval()' is used in a scope, it might introduce new
     * bindings in this or its parent scopes.
     * All those scopes are considered 'dynamic'.
     * @member {boolean} Scope#dynamic
     */

    this.dynamic = this.type === "global" || this.type === "with";
    /**
     * A reference to the scope-defining syntax node.
     * @member {espree.Node} Scope#block
     */

    this.block = block;
    /**
    * The {@link Reference|references} that are not resolved with this scope.
    * @member {Reference[]} Scope#through
    */

    this.through = [];
    /**
    * The scoped {@link Variable}s of this scope. In the case of a
    * 'function' scope this includes the automatic argument <em>arguments</em> as
    * its first element, as well as all further formal arguments.
    * @member {Variable[]} Scope#variables
    */

    this.variables = [];
    /**
    * Any variable {@link Reference|reference} found in this scope. This
    * includes occurrences of local variables as well as variables from
    * parent scopes (including the global scope). For local variables
    * this also includes defining occurrences (like in a 'var' statement).
    * In a 'function' scope this does not include the occurrences of the
    * formal parameter in the parameter list.
    * @member {Reference[]} Scope#references
    */

    this.references = [];
    /**
    * For 'global' and 'function' scopes, this is a self-reference. For
    * other scope types this is the <em>variableScope</em> value of the
    * parent scope.
    * @member {Scope} Scope#variableScope
    */

    this.variableScope = this.type === "global" || this.type === "function" || this.type === "module" ? this : upperScope.variableScope;
    /**
    * Whether this scope is created by a FunctionExpression.
    * @member {boolean} Scope#functionExpressionScope
    */

    this.functionExpressionScope = false;
    /**
    * Whether this is a scope that contains an 'eval()' invocation.
    * @member {boolean} Scope#directCallToEvalScope
    */

    this.directCallToEvalScope = false;
    /**
    * @member {boolean} Scope#thisFound
    */

    this.thisFound = false;
    this.__left = [];
    /**
    * Reference to the parent {@link Scope|scope}.
    * @member {Scope} Scope#upper
    */

    this.upper = upperScope;
    /**
    * Whether 'use strict' is in effect in this scope.
    * @member {boolean} Scope#isStrict
    */

    this.isStrict = isStrictScope(this, block, isMethodDefinition, scopeManager.__useDirective());
    /**
    * List of nested {@link Scope}s.
    * @member {Scope[]} Scope#childScopes
    */

    this.childScopes = [];

    if (this.upper) {
      this.upper.childScopes.push(this);
    }

    this.__declaredVariables = scopeManager.__declaredVariables;
    registerScope(scopeManager, this);
  }

  _createClass(Scope, [{
    key: "__shouldStaticallyClose",
    value: function __shouldStaticallyClose(scopeManager) {
      return !this.dynamic || scopeManager.__isOptimistic();
    }
  }, {
    key: "__shouldStaticallyCloseForGlobal",
    value: function __shouldStaticallyCloseForGlobal(ref) {
      // On global scope, let/const/class declarations should be resolved statically.
      var name = ref.identifier.name;

      if (!this.set.has(name)) {
        return false;
      }

      var variable = this.set.get(name);
      var defs = variable.defs;
      return defs.length > 0 && defs.every(shouldBeStatically);
    }
  }, {
    key: "__staticCloseRef",
    value: function __staticCloseRef(ref) {
      if (!this.__resolve(ref)) {
        this.__delegateToUpperScope(ref);
      }
    }
  }, {
    key: "__dynamicCloseRef",
    value: function __dynamicCloseRef(ref) {
      // notify all names are through to global
      var current = this;

      do {
        current.through.push(ref);
        current = current.upper;
      } while (current);
    }
  }, {
    key: "__globalCloseRef",
    value: function __globalCloseRef(ref) {
      // let/const/class declarations should be resolved statically.
      // others should be resolved dynamically.
      if (this.__shouldStaticallyCloseForGlobal(ref)) {
        this.__staticCloseRef(ref);
      } else {
        this.__dynamicCloseRef(ref);
      }
    }
  }, {
    key: "__close",
    value: function __close(scopeManager) {
      var closeRef;

      if (this.__shouldStaticallyClose(scopeManager)) {
        closeRef = this.__staticCloseRef;
      } else if (this.type !== "global") {
        closeRef = this.__dynamicCloseRef;
      } else {
        closeRef = this.__globalCloseRef;
      } // Try Resolving all references in this scope.


      for (var i = 0, iz = this.__left.length; i < iz; ++i) {
        var ref = this.__left[i];
        closeRef.call(this, ref);
      }

      this.__left = null;
      return this.upper;
    } // To override by function scopes.
    // References in default parameters isn't resolved to variables which are in their function body.

  }, {
    key: "__isValidResolution",
    value: function __isValidResolution(ref, variable) {
      // eslint-disable-line class-methods-use-this, no-unused-vars
      return true;
    }
  }, {
    key: "__resolve",
    value: function __resolve(ref) {
      var name = ref.identifier.name;

      if (!this.set.has(name)) {
        return false;
      }

      var variable = this.set.get(name);

      if (!this.__isValidResolution(ref, variable)) {
        return false;
      }

      variable.references.push(ref);
      variable.stack = variable.stack && ref.from.variableScope === this.variableScope;

      if (ref.tainted) {
        variable.tainted = true;
        this.taints.set(variable.name, true);
      }

      ref.resolved = variable;
      return true;
    }
  }, {
    key: "__delegateToUpperScope",
    value: function __delegateToUpperScope(ref) {
      if (this.upper) {
        this.upper.__left.push(ref);
      }

      this.through.push(ref);
    }
  }, {
    key: "__addDeclaredVariablesOfNode",
    value: function __addDeclaredVariablesOfNode(variable, node) {
      if (node === null || node === undefined) {
        return;
      }

      var variables = this.__declaredVariables.get(node);

      if (variables === null || variables === undefined) {
        variables = [];

        this.__declaredVariables.set(node, variables);
      }

      if (variables.indexOf(variable) === -1) {
        variables.push(variable);
      }
    }
  }, {
    key: "__defineGeneric",
    value: function __defineGeneric(name, set, variables, node, def) {
      var variable;
      variable = set.get(name);

      if (!variable) {
        variable = new Variable(name, this);
        set.set(name, variable);
        variables.push(variable);
      }

      if (def) {
        variable.defs.push(def);

        this.__addDeclaredVariablesOfNode(variable, def.node);

        this.__addDeclaredVariablesOfNode(variable, def.parent);
      }

      if (node) {
        variable.identifiers.push(node);
      }
    }
  }, {
    key: "__define",
    value: function __define(node, def) {
      if (node && node.type === Syntax.Identifier) {
        this.__defineGeneric(node.name, this.set, this.variables, node, def);
      }
    }
  }, {
    key: "__referencing",
    value: function __referencing(node, assign, writeExpr, maybeImplicitGlobal, partial, init) {
      // because Array element may be null
      if (!node || node.type !== Syntax.Identifier) {
        return;
      } // Specially handle like `this`.


      if (node.name === "super") {
        return;
      }

      var ref = new Reference(node, this, assign || Reference.READ, writeExpr, maybeImplicitGlobal, !!partial, !!init);
      this.references.push(ref);

      this.__left.push(ref);
    }
  }, {
    key: "__detectEval",
    value: function __detectEval() {
      var current = this;
      this.directCallToEvalScope = true;

      do {
        current.dynamic = true;
        current = current.upper;
      } while (current);
    }
  }, {
    key: "__detectThis",
    value: function __detectThis() {
      this.thisFound = true;
    }
  }, {
    key: "__isClosed",
    value: function __isClosed() {
      return this.__left === null;
    }
    /**
     * returns resolved {Reference}
     * @method Scope#resolve
     * @param {Espree.Identifier} ident - identifier to be resolved.
     * @returns {Reference} reference
     */

  }, {
    key: "resolve",
    value: function resolve(ident) {
      var ref, i, iz;
      assert(this.__isClosed(), "Scope should be closed.");
      assert(ident.type === Syntax.Identifier, "Target should be identifier.");

      for (i = 0, iz = this.references.length; i < iz; ++i) {
        ref = this.references[i];

        if (ref.identifier === ident) {
          return ref;
        }
      }

      return null;
    }
    /**
     * returns this scope is static
     * @method Scope#isStatic
     * @returns {boolean} static
     */

  }, {
    key: "isStatic",
    value: function isStatic() {
      return !this.dynamic;
    }
    /**
     * returns this scope has materialized arguments
     * @method Scope#isArgumentsMaterialized
     * @returns {boolean} arguemnts materialized
     */

  }, {
    key: "isArgumentsMaterialized",
    value: function isArgumentsMaterialized() {
      // eslint-disable-line class-methods-use-this
      return true;
    }
    /**
     * returns this scope has materialized `this` reference
     * @method Scope#isThisMaterialized
     * @returns {boolean} this materialized
     */

  }, {
    key: "isThisMaterialized",
    value: function isThisMaterialized() {
      // eslint-disable-line class-methods-use-this
      return true;
    }
  }, {
    key: "isUsedName",
    value: function isUsedName(name) {
      if (this.set.has(name)) {
        return true;
      }

      for (var i = 0, iz = this.through.length; i < iz; ++i) {
        if (this.through[i].identifier.name === name) {
          return true;
        }
      }

      return false;
    }
  }]);

  return Scope;
}();

var GlobalScope =
/*#__PURE__*/
function (_Scope) {
  _inherits(GlobalScope, _Scope);

  function GlobalScope(scopeManager, block) {
    var _this;

    _classCallCheck(this, GlobalScope);

    _this = _possibleConstructorReturn(this, _getPrototypeOf(GlobalScope).call(this, scopeManager, "global", null, block, false));
    _this.implicit = {
      set: new Map(),
      variables: [],

      /**
      * List of {@link Reference}s that are left to be resolved (i.e. which
      * need to be linked to the variable they refer to).
      * @member {Reference[]} Scope#implicit#left
      */
      left: []
    };
    return _this;
  }

  _createClass(GlobalScope, [{
    key: "__close",
    value: function __close(scopeManager) {
      var implicit = [];

      for (var i = 0, iz = this.__left.length; i < iz; ++i) {
        var ref = this.__left[i];

        if (ref.__maybeImplicitGlobal && !this.set.has(ref.identifier.name)) {
          implicit.push(ref.__maybeImplicitGlobal);
        }
      } // create an implicit global variable from assignment expression


      for (var _i2 = 0, _iz2 = implicit.length; _i2 < _iz2; ++_i2) {
        var info = implicit[_i2];

        this.__defineImplicit(info.pattern, new Definition(Variable.ImplicitGlobalVariable, info.pattern, info.node, null, null, null));
      }

      this.implicit.left = this.__left;
      return _get(_getPrototypeOf(GlobalScope.prototype), "__close", this).call(this, scopeManager);
    }
  }, {
    key: "__defineImplicit",
    value: function __defineImplicit(node, def) {
      if (node && node.type === Syntax.Identifier) {
        this.__defineGeneric(node.name, this.implicit.set, this.implicit.variables, node, def);
      }
    }
  }]);

  return GlobalScope;
}(Scope);

var ModuleScope =
/*#__PURE__*/
function (_Scope2) {
  _inherits(ModuleScope, _Scope2);

  function ModuleScope(scopeManager, upperScope, block) {
    _classCallCheck(this, ModuleScope);

    return _possibleConstructorReturn(this, _getPrototypeOf(ModuleScope).call(this, scopeManager, "module", upperScope, block, false));
  }

  return ModuleScope;
}(Scope);

var FunctionExpressionNameScope =
/*#__PURE__*/
function (_Scope3) {
  _inherits(FunctionExpressionNameScope, _Scope3);

  function FunctionExpressionNameScope(scopeManager, upperScope, block) {
    var _this2;

    _classCallCheck(this, FunctionExpressionNameScope);

    _this2 = _possibleConstructorReturn(this, _getPrototypeOf(FunctionExpressionNameScope).call(this, scopeManager, "function-expression-name", upperScope, block, false));

    _this2.__define(block.id, new Definition(Variable.FunctionName, block.id, block, null, null, null));

    _this2.functionExpressionScope = true;
    return _this2;
  }

  return FunctionExpressionNameScope;
}(Scope);

var CatchScope =
/*#__PURE__*/
function (_Scope4) {
  _inherits(CatchScope, _Scope4);

  function CatchScope(scopeManager, upperScope, block) {
    _classCallCheck(this, CatchScope);

    return _possibleConstructorReturn(this, _getPrototypeOf(CatchScope).call(this, scopeManager, "catch", upperScope, block, false));
  }

  return CatchScope;
}(Scope);

var WithScope =
/*#__PURE__*/
function (_Scope5) {
  _inherits(WithScope, _Scope5);

  function WithScope(scopeManager, upperScope, block) {
    _classCallCheck(this, WithScope);

    return _possibleConstructorReturn(this, _getPrototypeOf(WithScope).call(this, scopeManager, "with", upperScope, block, false));
  }

  _createClass(WithScope, [{
    key: "__close",
    value: function __close(scopeManager) {
      if (this.__shouldStaticallyClose(scopeManager)) {
        return _get(_getPrototypeOf(WithScope.prototype), "__close", this).call(this, scopeManager);
      }

      for (var i = 0, iz = this.__left.length; i < iz; ++i) {
        var ref = this.__left[i];
        ref.tainted = true;

        this.__delegateToUpperScope(ref);
      }

      this.__left = null;
      return this.upper;
    }
  }]);

  return WithScope;
}(Scope);

var BlockScope =
/*#__PURE__*/
function (_Scope6) {
  _inherits(BlockScope, _Scope6);

  function BlockScope(scopeManager, upperScope, block) {
    _classCallCheck(this, BlockScope);

    return _possibleConstructorReturn(this, _getPrototypeOf(BlockScope).call(this, scopeManager, "block", upperScope, block, false));
  }

  return BlockScope;
}(Scope);

var SwitchScope =
/*#__PURE__*/
function (_Scope7) {
  _inherits(SwitchScope, _Scope7);

  function SwitchScope(scopeManager, upperScope, block) {
    _classCallCheck(this, SwitchScope);

    return _possibleConstructorReturn(this, _getPrototypeOf(SwitchScope).call(this, scopeManager, "switch", upperScope, block, false));
  }

  return SwitchScope;
}(Scope);

var FunctionScope =
/*#__PURE__*/
function (_Scope8) {
  _inherits(FunctionScope, _Scope8);

  function FunctionScope(scopeManager, upperScope, block, isMethodDefinition) {
    var _this3;

    _classCallCheck(this, FunctionScope);

    _this3 = _possibleConstructorReturn(this, _getPrototypeOf(FunctionScope).call(this, scopeManager, "function", upperScope, block, isMethodDefinition)); // section 9.2.13, FunctionDeclarationInstantiation.
    // NOTE Arrow functions never have an arguments objects.

    if (_this3.block.type !== Syntax.ArrowFunctionExpression) {
      _this3.__defineArguments();
    }

    return _this3;
  }

  _createClass(FunctionScope, [{
    key: "isArgumentsMaterialized",
    value: function isArgumentsMaterialized() {
      // TODO(Constellation)
      // We can more aggressive on this condition like this.
      //
      // function t() {
      //     // arguments of t is always hidden.
      //     function arguments() {
      //     }
      // }
      if (this.block.type === Syntax.ArrowFunctionExpression) {
        return false;
      }

      if (!this.isStatic()) {
        return true;
      }

      var variable = this.set.get("arguments");
      assert(variable, "Always have arguments variable.");
      return variable.tainted || variable.references.length !== 0;
    }
  }, {
    key: "isThisMaterialized",
    value: function isThisMaterialized() {
      if (!this.isStatic()) {
        return true;
      }

      return this.thisFound;
    }
  }, {
    key: "__defineArguments",
    value: function __defineArguments() {
      this.__defineGeneric("arguments", this.set, this.variables, null, null);

      this.taints.set("arguments", true);
    } // References in default parameters isn't resolved to variables which are in their function body.
    //     const x = 1
    //     function f(a = x) { // This `x` is resolved to the `x` in the outer scope.
    //         const x = 2
    //         console.log(a)
    //     }

  }, {
    key: "__isValidResolution",
    value: function __isValidResolution(ref, variable) {
      // If `options.nodejsScope` is true, `this.block` becomes a Program node.
      if (this.block.type === "Program") {
        return true;
      }

      var bodyStart = this.block.body.range[0]; // It's invalid resolution in the following case:

      return !(variable.scope === this && ref.identifier.range[0] < bodyStart && // the reference is in the parameter part.
      variable.defs.every(function (d) {
        return d.name.range[0] >= bodyStart;
      }) // the variable is in the body.
      );
    }
  }]);

  return FunctionScope;
}(Scope);

var ForScope =
/*#__PURE__*/
function (_Scope9) {
  _inherits(ForScope, _Scope9);

  function ForScope(scopeManager, upperScope, block) {
    _classCallCheck(this, ForScope);

    return _possibleConstructorReturn(this, _getPrototypeOf(ForScope).call(this, scopeManager, "for", upperScope, block, false));
  }

  return ForScope;
}(Scope);

var ClassScope =
/*#__PURE__*/
function (_Scope10) {
  _inherits(ClassScope, _Scope10);

  function ClassScope(scopeManager, upperScope, block) {
    _classCallCheck(this, ClassScope);

    return _possibleConstructorReturn(this, _getPrototypeOf(ClassScope).call(this, scopeManager, "class", upperScope, block, false));
  }

  return ClassScope;
}(Scope);

module.exports = {
  Scope: Scope,
  GlobalScope: GlobalScope,
  ModuleScope: ModuleScope,
  FunctionExpressionNameScope: FunctionExpressionNameScope,
  CatchScope: CatchScope,
  WithScope: WithScope,
  BlockScope: BlockScope,
  SwitchScope: SwitchScope,
  FunctionScope: FunctionScope,
  ForScope: ForScope,
  ClassScope: ClassScope
};
/* vim: set sw=4 ts=4 et tw=80 : */
