// clipboardy/lib/dist/windows.dev.js
'use strict';

var path = require('path');

var execa = require('execa');

var arch = require('arch'); // Binaries from: https://github.com/sindresorhus/win-clipboard


var windowBinaryPath = arch() === 'x64' ? path.join(__dirname, '../fallbacks/windows/clipboard_x86_64.exe') : path.join(__dirname, '../fallbacks/windows/clipboard_i686.exe');
module.exports = {
  copy: function copy(options) {
    return regeneratorRuntime.async(function copy$(_context) {
      while (1) {
        switch (_context.prev = _context.next) {
          case 0:
            return _context.abrupt("return", execa(windowBinaryPath, ['--copy'], options));

          case 1:
          case "end":
            return _context.stop();
        }
      }
    });
  },
  paste: function paste(options) {
    return regeneratorRuntime.async(function paste$(_context2) {
      while (1) {
        switch (_context2.prev = _context2.next) {
          case 0:
            return _context2.abrupt("return", execa.stdout(windowBinaryPath, ['--paste'], options));

          case 1:
          case "end":
            return _context2.stop();
        }
      }
    });
  },
  copySync: function copySync(options) {
    return execa.sync(windowBinaryPath, ['--copy'], options);
  },
  pasteSync: function pasteSync(options) {
    return execa.sync(windowBinaryPath, ['--paste'], options);
  }
};
