// @vue/component-compiler-utils/lib/styleProcessors/dist/index.js
"use strict";
exports.__esModule = true;
exports.processors = void 0;
var merge = require('merge-source-map');
// .scss/.sass processor
var scss = {
    render: function (source, map, options) {
        var nodeSass = require('sass');
        var finalOptions = Object.assign({}, options, {
            data: source,
            file: options.filename,
            outFile: options.filename,
            sourceMap: !!map
        });
        try {
            var result = nodeSass.renderSync(finalOptions);
            if (map) {
                return {
                    code: result.css.toString(),
                    map: merge(map, JSON.parse(result.map.toString())),
                    errors: []
                };
            }
            return { code: result.css.toString(), errors: [] };
        }
        catch (e) {
            return { code: '', errors: [e] };
        }
    }
};
var sass = {
    render: function (source, map, options) {
        return scss.render(source, map, Object.assign({}, options, { indentedSyntax: true }));
    }
};
// .less
var less = {
    render: function (source, map, options) {
        var nodeLess = require('less');
        var result;
        var error = null;
        nodeLess.render(source, Object.assign({}, options, { syncImport: true }), function (err, output) {
            error = err;
            result = output;
        });
        if (error)
            return { code: '', errors: [error] };
        if (map) {
            return {
                code: result.css.toString(),
                map: merge(map, result.map),
                errors: []
            };
        }
        return { code: result.css.toString(), errors: [] };
    }
};
// .styl
var styl = {
    render: function (source, map, options) {
        var nodeStylus = require('stylus');
        try {
            var ref_1 = nodeStylus(source);
            Object.keys(options).forEach(function (key) { return ref_1.set(key, options[key]); });
            if (map)
                ref_1.set('sourcemap', { inline: false, comment: false });
            var result = ref_1.render();
            if (map) {
                return {
                    code: result,
                    map: merge(map, ref_1.sourcemap),
                    errors: []
                };
            }
            return { code: result, errors: [] };
        }
        catch (e) {
            return { code: '', errors: [e] };
        }
    }
};
exports.processors = {
    less: less,
    sass: sass,
    scss: scss,
    styl: styl,
    stylus: styl
};
