// multicast-dns-service-types/dist/index.dev.js
"use strict";

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

var prefix = function prefix(name) {
  return '_' + name;
};

var defined = function defined(name) {
  return name;
};

exports.stringify = function (data) {
  if (_typeof(data) === 'object' && data && data.name) return exports.stringify(data.name, data.protocol, data.subtypes);
  return Array.prototype.concat.apply([], arguments).filter(defined).map(prefix).join('.');
};

exports.parse = function (str) {
  var parts = str.split('.');

  for (var i = 0; i < parts.length; i++) {
    if (parts[i][0] !== '_') continue;
    parts[i] = parts[i].slice(1);
  }

  return {
    name: parts.shift(),
    protocol: parts.shift() || null,
    subtypes: parts
  };
};

exports.tcp = function (name) {
  return exports.stringify(name, 'tcp', Array.prototype.concat.apply([], Array.prototype.slice.call(arguments, 1)));
};

exports.udp = function (name) {
  return exports.stringify(name, 'udp', Array.prototype.concat.apply([], Array.prototype.slice.call(arguments, 1)));
};
