// thread-loader/dist/dist/worker.dev.js
'use strict';

var _fs = require('fs');

var _fs2 = _interopRequireDefault(_fs);

var _module = require('module');

var _module2 = _interopRequireDefault(_module);

var _loaderRunner = require('loader-runner');

var _loaderRunner2 = _interopRequireDefault(_loaderRunner);

var _queue = require('neo-async/queue');

var _queue2 = _interopRequireDefault(_queue);

var _readBuffer = require('./readBuffer');

var _readBuffer2 = _interopRequireDefault(_readBuffer);

function _interopRequireDefault(obj) {
  return obj && obj.__esModule ? obj : {
    "default": obj
  };
}

var writePipe = _fs2["default"].createWriteStream(null, {
  fd: 3
});
/* global require */

/* eslint-disable no-console */


var readPipe = _fs2["default"].createReadStream(null, {
  fd: 4
});

writePipe.on('finish', onTerminateWrite);
readPipe.on('end', onTerminateRead);
writePipe.on('close', onTerminateWrite);
readPipe.on('close', onTerminateRead);
readPipe.on('error', onError);
writePipe.on('error', onError);
var PARALLEL_JOBS = +process.argv[2] || 20;
var terminated = false;
var nextQuestionId = 0;
var callbackMap = Object.create(null);

function onError(error) {
  console.error(error);
}

function onTerminateRead() {
  terminateRead();
}

function onTerminateWrite() {
  terminateWrite();
}

function writePipeWrite() {
  if (!terminated) {
    writePipe.write.apply(writePipe, arguments);
  }
}

function writePipeCork() {
  if (!terminated) {
    writePipe.cork();
  }
}

function writePipeUncork() {
  if (!terminated) {
    writePipe.uncork();
  }
}

function terminateRead() {
  terminated = true;
  readPipe.removeAllListeners();
}

function terminateWrite() {
  terminated = true;
  writePipe.removeAllListeners();
}

function terminate() {
  terminateRead();
  terminateWrite();
}

function toErrorObj(err) {
  return {
    message: err.message,
    details: err.details,
    stack: err.stack,
    hideStack: err.hideStack
  };
}

function toNativeError(obj) {
  if (!obj) return null;
  var err = new Error(obj.message);
  err.details = obj.details;
  err.missing = obj.missing;
  return err;
}

function writeJson(data) {
  writePipeCork();
  process.nextTick(function () {
    writePipeUncork();
  });
  var lengthBuffer = Buffer.alloc(4);
  var messageBuffer = Buffer.from(JSON.stringify(data), 'utf-8');
  lengthBuffer.writeInt32BE(messageBuffer.length, 0);
  writePipeWrite(lengthBuffer);
  writePipeWrite(messageBuffer);
}

var queue = (0, _queue2["default"])(function (_ref, taskCallback) {
  var id = _ref.id,
      data = _ref.data;

  try {
    _loaderRunner2["default"].runLoaders({
      loaders: data.loaders,
      resource: data.resource,
      readResource: _fs2["default"].readFile.bind(_fs2["default"]),
      context: {
        version: 2,
        resolve: function resolve(context, request, callback) {
          callbackMap[nextQuestionId] = callback;
          writeJson({
            type: 'resolve',
            id: id,
            questionId: nextQuestionId,
            context: context,
            request: request
          });
          nextQuestionId += 1;
        },
        emitWarning: function emitWarning(warning) {
          writeJson({
            type: 'emitWarning',
            id: id,
            data: toErrorObj(warning)
          });
        },
        emitError: function emitError(error) {
          writeJson({
            type: 'emitError',
            id: id,
            data: toErrorObj(error)
          });
        },
        exec: function exec(code, filename) {
          var module = new _module2["default"](filename, undefined);
          module.paths = _module2["default"]._nodeModulePaths(undefined.context); // eslint-disable-line no-underscore-dangle

          module.filename = filename;

          module._compile(code, filename); // eslint-disable-line no-underscore-dangle


          return module.exports;
        },
        options: {
          context: data.optionsContext
        },
        webpack: true,
        'thread-loader': true,
        sourceMap: data.sourceMap,
        target: data.target,
        minimize: data.minimize,
        resourceQuery: data.resourceQuery
      }
    }, function (err, lrResult) {
      var result = lrResult.result,
          cacheable = lrResult.cacheable,
          fileDependencies = lrResult.fileDependencies,
          contextDependencies = lrResult.contextDependencies;
      var buffersToSend = [];
      var convertedResult = Array.isArray(result) && result.map(function (item) {
        var isBuffer = Buffer.isBuffer(item);

        if (isBuffer) {
          buffersToSend.push(item);
          return {
            buffer: true
          };
        }

        if (typeof item === 'string') {
          var stringBuffer = Buffer.from(item, 'utf-8');
          buffersToSend.push(stringBuffer);
          return {
            buffer: true,
            string: true
          };
        }

        return {
          data: item
        };
      });
      writeJson({
        type: 'job',
        id: id,
        error: err && toErrorObj(err),
        result: {
          result: convertedResult,
          cacheable: cacheable,
          fileDependencies: fileDependencies,
          contextDependencies: contextDependencies
        },
        data: buffersToSend.map(function (buffer) {
          return buffer.length;
        })
      });
      buffersToSend.forEach(function (buffer) {
        writePipeWrite(buffer);
      });
      setImmediate(taskCallback);
    });
  } catch (e) {
    writeJson({
      type: 'job',
      id: id,
      error: toErrorObj(e)
    });
    taskCallback();
  }
}, PARALLEL_JOBS);

function dispose() {
  terminate();
  queue.kill();
  process.exit(0);
}

function onMessage(message) {
  try {
    var type = message.type,
        id = message.id;

    switch (type) {
      case 'job':
        {
          queue.push(message);
          break;
        }

      case 'result':
        {
          var error = message.error,
              result = message.result;
          var callback = callbackMap[id];

          if (callback) {
            var nativeError = toNativeError(error);
            callback(nativeError, result);
          } else {
            console.error("Worker got unexpected result id ".concat(id));
          }

          delete callbackMap[id];
          break;
        }

      case 'warmup':
        {
          var requires = message.requires; // load modules into process

          requires.forEach(function (r) {
            return require(r);
          }); // eslint-disable-line import/no-dynamic-require, global-require

          break;
        }

      default:
        {
          console.error("Worker got unexpected job type ".concat(type));
          break;
        }
    }
  } catch (e) {
    console.error("Error in worker ".concat(e));
  }
}

function readNextMessage() {
  (0, _readBuffer2["default"])(readPipe, 4, function (lengthReadError, lengthBuffer) {
    if (lengthReadError) {
      console.error("Failed to communicate with main process (read length) ".concat(lengthReadError));
      return;
    }

    var length = lengthBuffer.length && lengthBuffer.readInt32BE(0);

    if (length === 0) {
      // worker should dispose and exit
      dispose();
      return;
    }

    (0, _readBuffer2["default"])(readPipe, length, function (messageError, messageBuffer) {
      if (terminated) {
        return;
      }

      if (messageError) {
        console.error("Failed to communicate with main process (read message) ".concat(messageError));
        return;
      }

      var messageString = messageBuffer.toString('utf-8');
      var message = JSON.parse(messageString);
      onMessage(message);
      setImmediate(function () {
        return readNextMessage();
      });
    });
  });
} // start reading messages from main process


readNextMessage();
