// globby/dist/gitignore.dev.js
'use strict';

var fs = require('fs');

var path = require('path');

var glob = require('glob');

var gitIgnore = require('ignore');

var pify = require('pify');

var slash = require('slash');

var globP = pify(glob);
var readFileP = pify(fs.readFile);

var mapGitIgnorePatternTo = function mapGitIgnorePatternTo(base) {
  return function (ignore) {
    if (ignore.startsWith('!')) {
      return '!' + path.posix.join(base, ignore.substr(1));
    }

    return path.posix.join(base, ignore);
  };
};

var parseGitIgnore = function parseGitIgnore(content, opts) {
  var base = slash(path.relative(opts.cwd, path.dirname(opts.fileName)));
  return content.split(/\r?\n/).filter(Boolean).filter(function (l) {
    return l.charAt(0) !== '#';
  }).map(mapGitIgnorePatternTo(base));
};

var reduceIgnore = function reduceIgnore(files) {
  return files.reduce(function (ignores, file) {
    ignores.add(parseGitIgnore(file.content, {
      cwd: file.cwd,
      fileName: file.filePath
    }));
    return ignores;
  }, gitIgnore());
};

var getIsIgnoredPredecate = function getIsIgnoredPredecate(ignores, cwd) {
  return function (p) {
    return ignores.ignores(slash(path.relative(cwd, p)));
  };
};

var getFile = function getFile(file, cwd) {
  var filePath = path.join(cwd, file);
  return readFileP(filePath, 'utf8').then(function (content) {
    return {
      content: content,
      cwd: cwd,
      filePath: filePath
    };
  });
};

var getFileSync = function getFileSync(file, cwd) {
  var filePath = path.join(cwd, file);
  var content = fs.readFileSync(filePath, 'utf8');
  return {
    content: content,
    cwd: cwd,
    filePath: filePath
  };
};

var normalizeOpts = function normalizeOpts(opts) {
  opts = opts || {};
  var ignore = opts.ignore || [];
  var cwd = opts.cwd || process.cwd();
  return {
    ignore: ignore,
    cwd: cwd
  };
};

module.exports = function (o) {
  var opts = normalizeOpts(o);
  return globP('**/.gitignore', {
    ignore: opts.ignore,
    cwd: opts.cwd
  }).then(function (paths) {
    return Promise.all(paths.map(function (file) {
      return getFile(file, opts.cwd);
    }));
  }).then(function (files) {
    return reduceIgnore(files);
  }).then(function (ignores) {
    return getIsIgnoredPredecate(ignores, opts.cwd);
  });
};

module.exports.sync = function (o) {
  var opts = normalizeOpts(o);
  var paths = glob.sync('**/.gitignore', {
    ignore: opts.ignore,
    cwd: opts.cwd
  });
  var files = paths.map(function (file) {
    return getFileSync(file, opts.cwd);
  });
  var ignores = reduceIgnore(files);
  return getIsIgnoredPredecate(ignores, opts.cwd);
};
