// file-loader/dist/dist/index.dev.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = loader;
exports.raw = void 0;

var _path = _interopRequireDefault(require("path"));

var _loaderUtils = _interopRequireDefault(require("loader-utils"));

var _schemaUtils = _interopRequireDefault(require("schema-utils"));

var _options = _interopRequireDefault(require("./options.json"));

function _interopRequireDefault(obj) {
  return obj && obj.__esModule ? obj : {
    "default": obj
  };
}

function loader(content) {
  var options = _loaderUtils["default"].getOptions(this) || {};
  (0, _schemaUtils["default"])(_options["default"], options, {
    name: 'File Loader',
    baseDataPath: 'options'
  });
  var context = options.context || this.rootContext;

  var url = _loaderUtils["default"].interpolateName(this, options.name || '[contenthash].[ext]', {
    context: context,
    content: content,
    regExp: options.regExp
  });

  var outputPath = url;

  if (options.outputPath) {
    if (typeof options.outputPath === 'function') {
      outputPath = options.outputPath(url, this.resourcePath, context);
    } else {
      outputPath = _path["default"].posix.join(options.outputPath, url);
    }
  }

  var publicPath = "__webpack_public_path__ + ".concat(JSON.stringify(outputPath));

  if (options.publicPath) {
    if (typeof options.publicPath === 'function') {
      publicPath = options.publicPath(url, this.resourcePath, context);
    } else {
      publicPath = "".concat(options.publicPath.endsWith('/') ? options.publicPath : "".concat(options.publicPath, "/")).concat(url);
    }

    publicPath = JSON.stringify(publicPath);
  }

  if (options.postTransformPublicPath) {
    publicPath = options.postTransformPublicPath(publicPath);
  }

  if (typeof options.emitFile === 'undefined' || options.emitFile) {
    this.emitFile(outputPath, content);
  }

  return "".concat(options.esModules ? 'export default' : 'module.exports =', " ").concat(publicPath, ";");
}

var raw = true;
exports.raw = raw;
