// dir-glob/dist/index.dev.js
'use strict';

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

var path = require('path');

var pathType = require('path-type');

var getExtensions = function getExtensions(extensions) {
  return extensions.length > 1 ? "{".concat(extensions.join(','), "}") : extensions[0];
};

var getPath = function getPath(filepath, cwd) {
  var pth = filepath[0] === '!' ? filepath.slice(1) : filepath;
  return path.isAbsolute(pth) ? pth : path.join(cwd, pth);
};

var addExtensions = function addExtensions(file, extensions) {
  if (path.extname(file)) {
    return "**/".concat(file);
  }

  return "**/".concat(file, ".").concat(getExtensions(extensions));
};

var getGlob = function getGlob(dir, opts) {
  if (opts.files && !Array.isArray(opts.files)) {
    throw new TypeError("Expected `files` to be of type `Array` but received type `".concat(_typeof(opts.files), "`"));
  }

  if (opts.extensions && !Array.isArray(opts.extensions)) {
    throw new TypeError("Expected `extensions` to be of type `Array` but received type `".concat(_typeof(opts.extensions), "`"));
  }

  if (opts.files && opts.extensions) {
    return opts.files.map(function (x) {
      return path.join(dir, addExtensions(x, opts.extensions));
    });
  }

  if (opts.files) {
    return opts.files.map(function (x) {
      return path.join(dir, "**/".concat(x));
    });
  }

  if (opts.extensions) {
    return [path.join(dir, "**/*.".concat(getExtensions(opts.extensions)))];
  }

  return [path.join(dir, '**')];
};

module.exports = function (input, opts) {
  opts = Object.assign({
    cwd: process.cwd()
  }, opts);

  if (typeof opts.cwd !== 'string') {
    return Promise.reject(new TypeError("Expected `cwd` to be of type `string` but received type `".concat(_typeof(opts.cwd), "`")));
  }

  return Promise.all([].concat(input).map(function (x) {
    return pathType.dir(getPath(x, opts.cwd)).then(function (isDir) {
      return isDir ? getGlob(x, opts) : x;
    });
  })).then(function (globs) {
    return [].concat.apply([], globs);
  });
};

module.exports.sync = function (input, opts) {
  opts = Object.assign({
    cwd: process.cwd()
  }, opts);

  if (typeof opts.cwd !== 'string') {
    throw new TypeError("Expected `cwd` to be of type `string` but received type `".concat(_typeof(opts.cwd), "`"));
  }

  var globs = [].concat(input).map(function (x) {
    return pathType.dirSync(getPath(x, opts.cwd)) ? getGlob(x, opts) : x;
  });
  return [].concat.apply([], globs);
};
