// webpack/lib/dist/ErrorHelpers.dev.js
/*
	MIT License http://www.opensource.org/licenses/mit-license.php
	Author Tobias Koppers @sokra
*/
"use strict";

var loaderFlag = "LOADER_EXECUTION";
var webpackOptionsFlag = "WEBPACK_OPTIONS";

exports.cutOffByFlag = function (stack, flag) {
  stack = stack.split("\n");

  for (var i = 0; i < stack.length; i++) {
    if (stack[i].includes(flag)) {
      stack.length = i;
    }
  }

  return stack.join("\n");
};

exports.cutOffLoaderExecution = function (stack) {
  return exports.cutOffByFlag(stack, loaderFlag);
};

exports.cutOffWebpackOptions = function (stack) {
  return exports.cutOffByFlag(stack, webpackOptionsFlag);
};

exports.cutOffMultilineMessage = function (stack, message) {
  stack = stack.split("\n");
  message = message.split("\n");
  return stack.reduce(function (acc, line, idx) {
    return line.includes(message[idx]) ? acc : acc.concat(line);
  }, []).join("\n");
};

exports.cutOffMessage = function (stack, message) {
  var nextLine = stack.indexOf("\n");

  if (nextLine === -1) {
    return stack === message ? "" : stack;
  } else {
    var firstLine = stack.substr(0, nextLine);
    return firstLine === message ? stack.substr(nextLine + 1) : stack;
  }
};

exports.cleanUp = function (stack, message) {
  stack = exports.cutOffLoaderExecution(stack);
  stack = exports.cutOffMessage(stack, message);
  return stack;
};

exports.cleanUpWebpackOptions = function (stack, message) {
  stack = exports.cutOffWebpackOptions(stack);
  stack = exports.cutOffMultilineMessage(stack, message);
  return stack;
};
