// is-color-stop/dist/index.dev.js
'use strict';

var isRGB = require('./lib/isRGB');

var isRGBA = require('./lib/isRGBA');

var isHSL = require('./lib/isHSL');

var isHSLA = require('./lib/isHSLA');

var isHex = require('./lib/isHex');

var isCSSColorName = require('./lib/isCSSColorName');

var isTransparent = require('./lib/isTransparent');

var isCSSLengthUnit = require('./lib/isCSSLengthUnit');

var isStop = require('./lib/isStop');

function isColor(colorStr) {
  var color = isRGB(colorStr) || isRGBA(colorStr) || isHSL(colorStr) || isHSLA(colorStr) || isHex(colorStr) || isCSSColorName(colorStr) || isTransparent(colorStr);
  return color;
}

module.exports = function isColorStop(color, stop) {
  return isColor(color) && isStop(stop);
};

module.exports.isColor = isColor;
module.exports.isRGB = isRGB;
module.exports.isRGBA = isRGBA;
module.exports.isHSL = isHSL;
module.exports.isHSLA = isHSLA;
module.exports.isHex = isHex;
module.exports.isCSSColorName = isCSSColorName;
module.exports.isTransparent = isTransparent;
module.exports.isCSSLengthUnit = isCSSLengthUnit;
