// @vue/component-compiler-utils/lib/templateCompilerModules/dist/assetUrl.js
"use strict";
// vue compiler module for transforming `<tag>:<attribute>` to `require`
exports.__esModule = true;
var utils_1 = require("./utils");
var defaultOptions = {
    audio: 'src',
    video: ['src', 'poster'],
    source: 'src',
    img: 'src',
    image: ['xlink:href', 'href'],
    use: ['xlink:href', 'href']
};
exports["default"] = (function (userOptions, transformAssetUrlsOption) {
    var options = userOptions
        ? Object.assign({}, defaultOptions, userOptions)
        : defaultOptions;
    return {
        postTransformNode: function (node) {
            transform(node, options, transformAssetUrlsOption);
        }
    };
});
function transform(node, options, transformAssetUrlsOption) {
    var _loop_1 = function (tag) {
        if ((tag === '*' || node.tag === tag) && node.attrs) {
            var attributes_1 = options[tag];
            if (typeof attributes_1 === 'string') {
                node.attrs.some(function (attr) {
                    return rewrite(attr, attributes_1, transformAssetUrlsOption);
                });
            }
            else if (Array.isArray(attributes_1)) {
                attributes_1.forEach(function (item) {
                    return node.attrs.some(function (attr) { return rewrite(attr, item, transformAssetUrlsOption); });
                });
            }
        }
    };
    for (var tag in options) {
        _loop_1(tag);
    }
}
function rewrite(attr, name, transformAssetUrlsOption) {
    if (attr.name === name) {
        var value = attr.value;
        // only transform static URLs
        if (value.charAt(0) === '"' && value.charAt(value.length - 1) === '"') {
            attr.value = utils_1.urlToRequire(value.slice(1, -1), transformAssetUrlsOption);
            return true;
        }
    }
    return false;
}
