// easy-stack/dist/stack.dev.js
"use strict";

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

var Stack =
/*#__PURE__*/
function () {
  function Stack() {
    _classCallCheck(this, Stack);

    this.stack = [];
    this.autoRun = true;
    this.running = false;
    this.stop = false;
  }

  _createClass(Stack, [{
    key: "clear",
    value: function clear() {
      this.stack = [];
      return this.stack;
    }
  }, {
    key: "contents",
    value: function contents(val) {
      if (val) {
        this.stack = val;
      }

      return this.stack;
    }
  }, {
    key: "add",
    value: function add() {
      var _this$stack;

      (_this$stack = this.stack).push.apply(_this$stack, arguments);

      if (!this.running && !this.stop && this.autoRun) {
        this.next();
      }
    }
  }, {
    key: "next",
    value: function next() {
      this.running = true;

      if (this.stack.length < 1 || this.stop) {
        this.running = false;
        return;
      }

      this.stack.pop().bind(this)();
    }
  }]);

  return Stack;
}();

module.exports = Stack;
