// css-loader/dist/dist/index.dev.js
"use strict";

function _toConsumableArray(arr) { return _arrayWithoutHoles(arr) || _iterableToArray(arr) || _nonIterableSpread(); }

function _nonIterableSpread() { throw new TypeError("Invalid attempt to spread non-iterable instance"); }

function _iterableToArray(iter) { if (Symbol.iterator in Object(iter) || Object.prototype.toString.call(iter) === "[object Arguments]") return Array.from(iter); }

function _arrayWithoutHoles(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = new Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = loader;

var _loaderUtils = require("loader-utils");

var _postcss = _interopRequireDefault(require("postcss"));

var _package = _interopRequireDefault(require("postcss/package.json"));

var _schemaUtils = _interopRequireDefault(require("schema-utils"));

var _semver = require("semver");

var _CssSyntaxError = _interopRequireDefault(require("./CssSyntaxError"));

var _Warning = _interopRequireDefault(require("./Warning"));

var _options = _interopRequireDefault(require("./options.json"));

var _plugins = require("./plugins");

var _utils = require("./utils");

function _interopRequireDefault(obj) {
  return obj && obj.__esModule ? obj : {
    "default": obj
  };
}
/*
  MIT License http://www.opensource.org/licenses/mit-license.php
  Author Tobias Koppers @sokra
*/


function loader(content, map, meta) {
  var _this = this;

  var options = (0, _loaderUtils.getOptions)(this) || {};
  (0, _schemaUtils["default"])(_options["default"], options, {
    name: 'CSS Loader',
    baseDataPath: 'options'
  });
  var callback = this.async();
  var sourceMap = options.sourceMap || false;
  var plugins = [];

  if ((0, _utils.shouldUseModulesPlugins)(options.modules, this.resourcePath)) {
    plugins.push.apply(plugins, _toConsumableArray((0, _utils.getModulesPlugins)(options, this)));
  }

  var exportType = options.onlyLocals ? 'locals' : 'full';
  var preRequester = (0, _utils.getPreRequester)(this);

  var urlHandler = function urlHandler(url) {
    return (0, _loaderUtils.stringifyRequest)(_this, preRequester(options.importLoaders) + url);
  };

  plugins.push((0, _plugins.icssParser)({
    urlHandler: urlHandler
  }));

  if (options["import"] !== false && exportType === 'full') {
    plugins.push((0, _plugins.importParser)({
      filter: (0, _utils.getFilter)(options["import"], this.resourcePath),
      urlHandler: urlHandler
    }));
  }

  if (options.url !== false && exportType === 'full') {
    plugins.push((0, _plugins.urlParser)({
      filter: (0, _utils.getFilter)(options.url, this.resourcePath, function (value) {
        return (0, _loaderUtils.isUrlRequest)(value);
      }),
      urlHandler: function urlHandler(url) {
        return (0, _loaderUtils.stringifyRequest)(_this, url);
      }
    }));
  } // Reuse CSS AST (PostCSS AST e.g 'postcss-loader') to avoid reparsing


  if (meta) {
    var ast = meta.ast;

    if (ast && ast.type === 'postcss' && (0, _semver.satisfies)(ast.version, "^".concat(_package["default"].version))) {
      // eslint-disable-next-line no-param-reassign
      content = ast.root;
    }
  }

  (0, _postcss["default"])(plugins).process(content, {
    from: this.resourcePath,
    to: this.resourcePath,
    map: options.sourceMap ? {
      // Some loaders (example `"postcss-loader": "1.x.x"`) always generates source map, we should remove it
      prev: sourceMap && map ? (0, _utils.normalizeSourceMap)(map) : null,
      inline: false,
      annotation: false
    } : false
  }).then(function (result) {
    var _iteratorNormalCompletion = true;
    var _didIteratorError = false;
    var _iteratorError = undefined;

    try {
      for (var _iterator = result.warnings()[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
        var warning = _step.value;

        _this.emitWarning(new _Warning["default"](warning));
      }
    } catch (err) {
      _didIteratorError = true;
      _iteratorError = err;
    } finally {
      try {
        if (!_iteratorNormalCompletion && _iterator["return"] != null) {
          _iterator["return"]();
        }
      } finally {
        if (_didIteratorError) {
          throw _iteratorError;
        }
      }
    }

    var imports = [];
    var apiImports = [];
    var urlReplacements = [];
    var icssReplacements = [];
    var exports = [];
    var _iteratorNormalCompletion2 = true;
    var _didIteratorError2 = false;
    var _iteratorError2 = undefined;

    try {
      for (var _iterator2 = result.messages[Symbol.iterator](), _step2; !(_iteratorNormalCompletion2 = (_step2 = _iterator2.next()).done); _iteratorNormalCompletion2 = true) {
        var message = _step2.value;

        // eslint-disable-next-line default-case
        switch (message.type) {
          case 'import':
            imports.push(message.value);
            break;

          case 'api-import':
            apiImports.push(message.value);
            break;

          case 'url-replacement':
            urlReplacements.push(message.value);
            break;

          case 'icss-replacement':
            icssReplacements.push(message.value);
            break;

          case 'export':
            exports.push(message.value);
            break;
        }
      }
    } catch (err) {
      _didIteratorError2 = true;
      _iteratorError2 = err;
    } finally {
      try {
        if (!_iteratorNormalCompletion2 && _iterator2["return"] != null) {
          _iterator2["return"]();
        }
      } finally {
        if (_didIteratorError2) {
          throw _iteratorError2;
        }
      }
    }

    var localsConvention = options.localsConvention;
    var esModule = typeof options.esModule !== 'undefined' ? options.esModule : false;
    var importCode = (0, _utils.getImportCode)(_this, exportType, imports, esModule);
    var moduleCode = (0, _utils.getModuleCode)(result, exportType, sourceMap, apiImports, urlReplacements, icssReplacements, esModule);
    var exportCode = (0, _utils.getExportCode)(exports, exportType, localsConvention, icssReplacements, esModule);
    return callback(null, "".concat(importCode).concat(moduleCode).concat(exportCode));
  })["catch"](function (error) {
    if (error.file) {
      _this.addDependency(error.file);
    }

    callback(error.name === 'CssSyntaxError' ? new _CssSyntaxError["default"](error) : error);
  });
}
