// webpack-log/src/dist/index.dev.js
'use strict';
/* global window: true */

/* eslint-disable
  no-shadow,
  no-param-reassign,
  space-before-function-paren
*/

var uuid = require('uuid/v4');

var colors = require('ansi-colors');

var loglevel = require('./loglevel');

var symbols = {
  trace: colors.grey('₸'),
  debug: colors.cyan('➤'),
  info: colors.blue(colors.symbols.info),
  warn: colors.yellow(colors.symbols.warning),
  error: colors.red(colors.symbols.cross)
};
var defaults = {
  name: '<unknown>',
  level: 'info',
  unique: true
};
var prefix = {
  level: function level(options) {
    return symbols[options.level];
  },
  template: "{{level}} ".concat(colors.gray('｢{{name}}｣'), ": ")
};

function log(options) {
  var opts = Object.assign({}, defaults, options);
  var id = options.id;
  opts.prefix = Object.assign({}, prefix, options.prefix);
  delete opts.id;
  Object.defineProperty(opts, 'id', {
    get: function get() {
      if (!id) {
        return this.name + (opts.unique ? "-".concat(uuid()) : '');
      }

      return id;
    }
  });

  if (opts.timestamp) {
    opts.prefix.template = "[{{time}}] ".concat(opts.prefix.template);
  }

  var log = loglevel.getLogger(opts);

  if (!Object.prototype.hasOwnProperty.call(log, 'id')) {
    Object.defineProperty(log, 'id', {
      get: function get() {
        return opts.id;
      }
    });
  }

  return log;
}

module.exports = log; // NOTE: this is exported so that consumers of webpack-log can use the same
// version of ansi-colors to decorate log messages without incurring additional
// dependency overhead

module.exports.colors = colors; // NOTE: This is an undocumented function solely for the purpose of tests.
// Do not use this method in production code. Using in production code
// may result in strange behavior.

module.exports.delLogger = function delLogger(name) {
  delete loglevel.loggers[name];
};

module.exports.factories = loglevel.factories;
