// tapable/lib/dist/Hook.dev.js
/*
	MIT License http://www.opensource.org/licenses/mit-license.php
	Author Tobias Koppers @sokra
*/
"use strict";

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

var Hook =
/*#__PURE__*/
function () {
  function Hook(args) {
    _classCallCheck(this, Hook);

    if (!Array.isArray(args)) args = [];
    this._args = args;
    this.taps = [];
    this.interceptors = [];
    this.call = this._call;
    this.promise = this._promise;
    this.callAsync = this._callAsync;
    this._x = undefined;
  }

  _createClass(Hook, [{
    key: "compile",
    value: function compile(options) {
      throw new Error("Abstract: should be overriden");
    }
  }, {
    key: "_createCall",
    value: function _createCall(type) {
      return this.compile({
        taps: this.taps,
        interceptors: this.interceptors,
        args: this._args,
        type: type
      });
    }
  }, {
    key: "tap",
    value: function tap(options, fn) {
      if (typeof options === "string") options = {
        name: options
      };
      if (_typeof(options) !== "object" || options === null) throw new Error("Invalid arguments to tap(options: Object, fn: function)");
      options = Object.assign({
        type: "sync",
        fn: fn
      }, options);
      if (typeof options.name !== "string" || options.name === "") throw new Error("Missing name for tap");
      options = this._runRegisterInterceptors(options);

      this._insert(options);
    }
  }, {
    key: "tapAsync",
    value: function tapAsync(options, fn) {
      if (typeof options === "string") options = {
        name: options
      };
      if (_typeof(options) !== "object" || options === null) throw new Error("Invalid arguments to tapAsync(options: Object, fn: function)");
      options = Object.assign({
        type: "async",
        fn: fn
      }, options);
      if (typeof options.name !== "string" || options.name === "") throw new Error("Missing name for tapAsync");
      options = this._runRegisterInterceptors(options);

      this._insert(options);
    }
  }, {
    key: "tapPromise",
    value: function tapPromise(options, fn) {
      if (typeof options === "string") options = {
        name: options
      };
      if (_typeof(options) !== "object" || options === null) throw new Error("Invalid arguments to tapPromise(options: Object, fn: function)");
      options = Object.assign({
        type: "promise",
        fn: fn
      }, options);
      if (typeof options.name !== "string" || options.name === "") throw new Error("Missing name for tapPromise");
      options = this._runRegisterInterceptors(options);

      this._insert(options);
    }
  }, {
    key: "_runRegisterInterceptors",
    value: function _runRegisterInterceptors(options) {
      var _iteratorNormalCompletion = true;
      var _didIteratorError = false;
      var _iteratorError = undefined;

      try {
        for (var _iterator = this.interceptors[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
          var interceptor = _step.value;

          if (interceptor.register) {
            var newOptions = interceptor.register(options);
            if (newOptions !== undefined) options = newOptions;
          }
        }
      } catch (err) {
        _didIteratorError = true;
        _iteratorError = err;
      } finally {
        try {
          if (!_iteratorNormalCompletion && _iterator["return"] != null) {
            _iterator["return"]();
          }
        } finally {
          if (_didIteratorError) {
            throw _iteratorError;
          }
        }
      }

      return options;
    }
  }, {
    key: "withOptions",
    value: function withOptions(options) {
      var mergeOptions = function mergeOptions(opt) {
        return Object.assign({}, options, typeof opt === "string" ? {
          name: opt
        } : opt);
      }; // Prevent creating endless prototype chains


      options = Object.assign({}, options, this._withOptions);
      var base = this._withOptionsBase || this;
      var newHook = Object.create(base);
      newHook.tapAsync = function (opt, fn) {
        return base.tapAsync(mergeOptions(opt), fn);
      }, newHook.tap = function (opt, fn) {
        return base.tap(mergeOptions(opt), fn);
      };

      newHook.tapPromise = function (opt, fn) {
        return base.tapPromise(mergeOptions(opt), fn);
      };

      newHook._withOptions = options;
      newHook._withOptionsBase = base;
      return newHook;
    }
  }, {
    key: "isUsed",
    value: function isUsed() {
      return this.taps.length > 0 || this.interceptors.length > 0;
    }
  }, {
    key: "intercept",
    value: function intercept(interceptor) {
      this._resetCompilation();

      this.interceptors.push(Object.assign({}, interceptor));

      if (interceptor.register) {
        for (var i = 0; i < this.taps.length; i++) {
          this.taps[i] = interceptor.register(this.taps[i]);
        }
      }
    }
  }, {
    key: "_resetCompilation",
    value: function _resetCompilation() {
      this.call = this._call;
      this.callAsync = this._callAsync;
      this.promise = this._promise;
    }
  }, {
    key: "_insert",
    value: function _insert(item) {
      this._resetCompilation();

      var before;
      if (typeof item.before === "string") before = new Set([item.before]);else if (Array.isArray(item.before)) {
        before = new Set(item.before);
      }
      var stage = 0;
      if (typeof item.stage === "number") stage = item.stage;
      var i = this.taps.length;

      while (i > 0) {
        i--;
        var x = this.taps[i];
        this.taps[i + 1] = x;
        var xStage = x.stage || 0;

        if (before) {
          if (before.has(x.name)) {
            before["delete"](x.name);
            continue;
          }

          if (before.size > 0) {
            continue;
          }
        }

        if (xStage > stage) {
          continue;
        }

        i++;
        break;
      }

      this.taps[i] = item;
    }
  }]);

  return Hook;
}();

function createCompileDelegate(name, type) {
  return function lazyCompileHook() {
    this[name] = this._createCall(type);
    return this[name].apply(this, arguments);
  };
}

Object.defineProperties(Hook.prototype, {
  _call: {
    value: createCompileDelegate("call", "sync"),
    configurable: true,
    writable: true
  },
  _promise: {
    value: createCompileDelegate("promise", "promise"),
    configurable: true,
    writable: true
  },
  _callAsync: {
    value: createCompileDelegate("callAsync", "async"),
    configurable: true,
    writable: true
  }
});
module.exports = Hook;
