// clipboardy/dist/browser.dev.js
/* eslint-env browser */
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.writeSync = exports.readSync = exports.read = exports.write = void 0;

var write = function write(text) {
  return regeneratorRuntime.async(function write$(_context) {
    while (1) {
      switch (_context.prev = _context.next) {
        case 0:
          _context.next = 2;
          return regeneratorRuntime.awrap(navigator.clipboard.writeText(text));

        case 2:
        case "end":
          return _context.stop();
      }
    }
  });
};

exports.write = write;

var read = function read() {
  return regeneratorRuntime.async(function read$(_context2) {
    while (1) {
      switch (_context2.prev = _context2.next) {
        case 0:
          return _context2.abrupt("return", navigator.clipboard.readText());

        case 1:
        case "end":
          return _context2.stop();
      }
    }
  });
};

exports.read = read;

var readSync = function readSync() {
  throw new Error('`.readSync()` is not supported in browsers!');
};

exports.readSync = readSync;

var writeSync = function writeSync() {
  throw new Error('`.writeSync()` is not supported in browsers!');
};

exports.writeSync = writeSync;
