// vue/src/server/webpack-plugin/dist/client.dev.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _util = require("./util");

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

var hash = require('hash-sum');

var uniq = require('lodash.uniq');

var VueSSRClientPlugin =
/*#__PURE__*/
function () {
  function VueSSRClientPlugin() {
    var options = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

    _classCallCheck(this, VueSSRClientPlugin);

    this.options = Object.assign({
      filename: 'vue-ssr-client-manifest.json'
    }, options);
  }

  _createClass(VueSSRClientPlugin, [{
    key: "apply",
    value: function apply(compiler) {
      var _this = this;

      (0, _util.onEmit)(compiler, 'vue-client-plugin', function (compilation, cb) {
        var stats = compilation.getStats().toJson();
        var allFiles = uniq(stats.assets.map(function (a) {
          return a.name;
        }));
        var initialFiles = uniq(Object.keys(stats.entrypoints).map(function (name) {
          return stats.entrypoints[name].assets;
        }).reduce(function (assets, all) {
          return all.concat(assets);
        }, []).filter(function (file) {
          return (0, _util.isJS)(file) || (0, _util.isCSS)(file);
        }));
        var asyncFiles = allFiles.filter(function (file) {
          return (0, _util.isJS)(file) || (0, _util.isCSS)(file);
        }).filter(function (file) {
          return initialFiles.indexOf(file) < 0;
        });
        var manifest = {
          publicPath: stats.publicPath,
          all: allFiles,
          initial: initialFiles,
          async: asyncFiles,
          modules: {
            /* [identifier: string]: Array<index: number> */
          }
        };
        var assetModules = stats.modules.filter(function (m) {
          return m.assets.length;
        });

        var fileToIndex = function fileToIndex(file) {
          return manifest.all.indexOf(file);
        };

        stats.modules.forEach(function (m) {
          // ignore modules duplicated in multiple chunks
          if (m.chunks.length === 1) {
            var cid = m.chunks[0];
            var chunk = stats.chunks.find(function (c) {
              return c.id === cid;
            });

            if (!chunk || !chunk.files) {
              return;
            }

            var id = m.identifier.replace(/\s\w+$/, ''); // remove appended hash

            var files = manifest.modules[hash(id)] = chunk.files.map(fileToIndex); // find all asset modules associated with the same chunk

            assetModules.forEach(function (m) {
              if (m.chunks.some(function (id) {
                return id === cid;
              })) {
                files.push.apply(files, m.assets.map(fileToIndex));
              }
            });
          }
        });
        var json = JSON.stringify(manifest, null, 2);
        compilation.assets[_this.options.filename] = {
          source: function source() {
            return json;
          },
          size: function size() {
            return json.length;
          }
        };
        cb();
      });
    }
  }]);

  return VueSSRClientPlugin;
}();

exports["default"] = VueSSRClientPlugin;
