// vue/src/server/webpack-plugin/dist/util.dev.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "isJS", {
  enumerable: true,
  get: function get() {
    return _util.isJS;
  }
});
Object.defineProperty(exports, "isCSS", {
  enumerable: true,
  get: function get() {
    return _util.isCSS;
  }
});
exports.onEmit = exports.validate = void 0;

var _util = require("../util");

var _require = require('chalk'),
    red = _require.red,
    yellow = _require.yellow;

var prefix = "[vue-server-renderer-webpack-plugin]";

var warn = exports.warn = function (msg) {
  return console.error(red("".concat(prefix, " ").concat(msg, "\n")));
};

var tip = exports.tip = function (msg) {
  return console.log(yellow("".concat(prefix, " ").concat(msg, "\n")));
};

var validate = function validate(compiler) {
  if (compiler.options.target !== 'node') {
    warn('webpack config `target` should be "node".');
  }

  if (compiler.options.output && compiler.options.output.libraryTarget !== 'commonjs2') {
    warn('webpack config `output.libraryTarget` should be "commonjs2".');
  }

  if (!compiler.options.externals) {
    tip('It is recommended to externalize dependencies in the server build for ' + 'better build performance.');
  }
};

exports.validate = validate;

var onEmit = function onEmit(compiler, name, hook) {
  if (compiler.hooks) {
    // Webpack >= 4.0.0
    compiler.hooks.emit.tapAsync(name, hook);
  } else {
    // Webpack < 4.0.0
    compiler.plugin('emit', hook);
  }
};

exports.onEmit = onEmit;
