// vue-loader-v16/dist/dist/stylePostLoader.dev.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

var qs = require("querystring");

var compiler_sfc_1 = require("@vue/compiler-sfc"); // This is a post loader that handles scoped CSS transforms.
// Injected right before css-loader by the global pitcher (../pitch.js)
// for any <style scoped> selection requests initiated from within vue files.


var StylePostLoader = function StylePostLoader(source, inMap) {
  var query = qs.parse(this.resourceQuery.slice(1));

  var _compiler_sfc_1$compi = compiler_sfc_1.compileStyle({
    source: source,
    filename: this.resourcePath,
    id: "data-v-".concat(query.id),
    map: inMap,
    scoped: !!query.scoped,
    vars: !!query.vars,
    trim: true
  }),
      code = _compiler_sfc_1$compi.code,
      map = _compiler_sfc_1$compi.map,
      errors = _compiler_sfc_1$compi.errors;

  if (errors.length) {
    this.callback(errors[0]);
  } else {
    this.callback(null, code, map);
  }
};

exports["default"] = StylePostLoader;
