// webpack/lib/dependencies/dist/LocalModule.dev.js
/*
	MIT License http://www.opensource.org/licenses/mit-license.php
	Author Tobias Koppers @sokra
*/
"use strict";

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

var LocalModule =
/*#__PURE__*/
function () {
  function LocalModule(module, name, idx) {
    _classCallCheck(this, LocalModule);

    this.module = module;
    this.name = name;
    this.idx = idx;
    this.used = false;
  }

  _createClass(LocalModule, [{
    key: "flagUsed",
    value: function flagUsed() {
      this.used = true;
    }
  }, {
    key: "variableName",
    value: function variableName() {
      return "__WEBPACK_LOCAL_MODULE_" + this.idx + "__";
    }
  }]);

  return LocalModule;
}();

module.exports = LocalModule;
