// @vue/cli-shared-utils/lib/dist/spinner.dev.js
"use strict";

var ora = require('ora');

var chalk = require('chalk');

var spinner = ora();
var lastMsg = null;
var isPaused = false;

exports.logWithSpinner = function (symbol, msg) {
  if (!msg) {
    msg = symbol;
    symbol = chalk.green('✔');
  }

  if (lastMsg) {
    spinner.stopAndPersist({
      symbol: lastMsg.symbol,
      text: lastMsg.text
    });
  }

  spinner.text = ' ' + msg;
  lastMsg = {
    symbol: symbol + ' ',
    text: msg
  };
  spinner.start();
};

exports.stopSpinner = function (persist) {
  if (!spinner.isSpinning) {
    return;
  }

  if (lastMsg && persist !== false) {
    spinner.stopAndPersist({
      symbol: lastMsg.symbol,
      text: lastMsg.text
    });
  } else {
    spinner.stop();
  }

  lastMsg = null;
};

exports.pauseSpinner = function () {
  if (spinner.isSpinning) {
    spinner.stop();
    isPaused = true;
  }
};

exports.resumeSpinner = function () {
  if (isPaused) {
    spinner.start();
    isPaused = false;
  }
};

exports.failSpinner = function (text) {
  spinner.fail(text);
}; // silent all logs except errors during tests and keep record


if (process.env.VUE_CLI_TEST) {
  require('./_silence')('spinner', exports);
}
